### DISCUSSION TITLE
"npx prisma migrate dev" fails -> Error: P1001: Can't reach database server at `localhost`:`3306`

### DISCUSSION THREAD

**Discussion by ManuWeb3** [prisma/prisma#21666] [Q&A] [answered]
### Question

In my basic Next.js13 project, I'm using Prisma ORM to interact with MySQL Server 8.2 instance set up locally at port 3306.

```schema.prisma``` & ```.env``` files pasted in the appropriate section.

Upon running the command:
```
npx prisma migrate dev
```

I get the error below:
```
Environment variables loaded from .env
Prisma schema loaded from prisma/schema.prisma
Datasource "db": MySQL database "DBSampleDB" at "127.0.0.1:3306"

Error: P1001: Can't reach database server at `127.0.0.1`:`3306`

Please make sure your database server is running at `127.0.0.1`:`3306`.
```

I ensured everytime the root password eneterd in ther URL is correct.

========

To ensure that **MySQL server is up and running and listening at port 3306** , I've checked bunch of things that prove that the server is indeed up at port 3306.

**Most evident** out of all those checks is the creation of a dummy database *DBSampleDB* using SQL Command Line and then creating a dummy table *Test_Table* with 2 columns (id, name) using 2 different IDEs: _DataGrip_ (JetBeans) and _DBeaver_ after **TestConnection** got **passed** using both the IDEs.

Snips below:

![image](https://github.com/prisma/prisma/assets/108497781/7b70915b-185b-4089-bf2b-6835efff4c4e)

![image](https://github.com/prisma/prisma/assets/108497781/2ae2832c-10ad-474b-8db0-91f2c9b90f36)

=============

Can this be a compatibility issue between versions of Prisma (_5.5.2_) and MySQL Server (_8.2_)?

I've exhausted all options suggested by ChatGPT and kapa.ai.

I've been stuck at this isuse for **4 days** now as I wasn't aware that this will boil down to Prisma though on the face of it, it seems Prisma but i did not want to jump the gun.

As I'm new to Web development, I had to rule out a whole bunch of things and learned debugging along the way (_and there were other issues that I got to know along the way,_ but the error still persists.)

By no means, I'm an expert and, hence, could have still missed to know other debugging options.

Please help as getting stuck for 4 days is a bit too much.

Thanks
Manu

### Expected behavior (optional)

As per the tutorial, it should have asked for the name of Migration as the next step and created a migration file under the folder _Prisma_ in the root directory of my codebase, hence, migrating the empty database to the schema defined in the ```schema.prisma```

### Information about Prisma Schema, Client Queries and Environment (optional)

```prisma
generator client {
  provider = "prisma-client-js"
}

datasource db {
  provider = "mysql"
  url      = env("DATABASE_URL")
}

model User {
  id        Int     @id @default(autoincrement())
  email     String  @db.VarChar(255)
  password  String  @db.VarChar(255)
  firstName String  @db.VarChar(255)  @map("first_name")
}
```

```.env``` file:

```env
DATABASE_URL="mysql://root:******@127.0.0.1:3306/DBSampleDB?connect_timeout=300"
```

```ts
Don't have any Prisma Client queries for now.
```

- OS:  Windows 11
- Database:  MySQL Server 8.2
- Node.js version:  v18.18.2

- Run `prisma -v` to see your Prisma version and paste it
5.5.2

**Comment by ManuWeb3** [comment]
Hi @ruheni @nurul3101 , though my first choice was not to tag and bother anyone from the Prisma team while posting my question here but as it's been around 17 hours and no reply, so thought of tagging.
I'm stuck at this error and can't proceed with anything unless it gets resolved or I get a workaround.

Thank you.

**Reply by janpio** [reply]
Today is Sunday. Most of my colleagues don't work today.

**Reply by ManuWeb3** [reply]
Hi @janpio , I'm sorry. I missed to take note of that.
Thanks for letting know.

**Reply by ManuWeb3** [reply]
Hi Prisma team, can someone please provide a fix for this?
I'm still stuck here.

**Reply by janpio** [reply]
There is nothing to fix. Something is off with that connection string if Prisma can not connect to it.

(For example I see that you use `DBSampleDB` for Prisma, but that Dbeaver shows `dbsampledb`.)

**Reply by ManuWeb3** [reply]
![image](https://github.com/prisma/prisma/assets/108497781/c2379b6b-4f86-4b83-ab71-6fcc40106ca4)

**1).** 
I always the gave the Database name ```DBSampleDB``` matching the exact case of alphabets in DataGrip IDE while setting it up.
Test Connection always passed.
Only under the hood, it shows the name in lower case: ```dbsampledb```. So, I don't see this an issue or the reason.

**2).** 
![image](https://github.com/prisma/prisma/assets/108497781/abae21c9-5443-4a67-b193-6d3c2556a23e)
Exactly the same setting with DBeaver also.

**3).** 
- I ensured the username is ```root``` and the password is correct while setting up MySQL Configurator and the connection url string.
- Tried switching default ```3306``` port to ```3309``` in Configurator and the string
- Tried with and without ```connect_timeout```
- Tried with and without the double quotes ```"  "``` around the url string while assigning it to ```.env``` variable ```DATABASE_URL``` in ```.env``` file
- Switched ```localhost``` to ```127.0.0.1``` and vice-versa.
- Turning on and off my Firewall and Anti-virus continuous scanning to see if they are blocking the port anyhow.
- uninstalled / re-installed MySQL and Prisma package.
- Changed Windows PowerShell Execution Policy to ```RemoteSigned``` from ```Restricted``` to ensure enough scripts run under the hood and are not blocked. However, this was done to install Chocolatey Package Manager during debugging.

In short, ran everything under the sun that I learned online, from ChatGPT, and kapa.ai in Prisma's Discord server but everytime the same issue persisted.

Then only, reached out to the Prisma team here.

**Reply by janpio** [reply]
Did you try using a lowercase database name in your Prisma connection string?

**Reply by ManuWeb3** [reply]
Yes, originally I named the DB in the url as ```nextapp``` and encountered this issue for the first time.

![image](https://github.com/prisma/prisma/assets/108497781/da1d6e94-2a74-4972-86a5-9884dbf191bc)

Then, during debugging, I named it differently for no specific reason though.

**Reply by janpio** [reply]
Can you register an account on planetscale.com and get a MySQL database there and try with that connection string?
Just to confirm and show that your Prisma is capable of connecting to MySQL databases.

**Reply by ManuWeb3** [reply]
Sure, I'll do it right away and update you.

However, I just re-tried with ```nextapp```, same error.
Then, I manually created the db: ```nextapp``` using MySQL's Command Line client.

![image](https://github.com/prisma/prisma/assets/108497781/2d70698b-3c08-46c9-8f97-71eed590b618)

DataGrip's _Test Connection_ passed but the migration did not.

![image](https://github.com/prisma/prisma/assets/108497781/4b9bb8a5-42a3-40bd-b186-0388211461e3)

**Reply by ManuWeb3** [reply]
Hi @janpio , I'm able to connect to MySQL database with PlanetScale via Prisma.

```npx prisma db push```

![prisma db push output](https://github.com/prisma/prisma/assets/108497781/61738df4-738c-4aae-9efe-1c6933c5ee2c)

It created 2 tables: ```User``` and ```Item``` in ```planetdb``` database in sync with Prisma schema.

![main_branch_2_tables](https://github.com/prisma/prisma/assets/108497781/a86799e9-09f6-4671-b0ee-8b15b832fabe)

```schema.prisma```

```TypeScript
datasource db {
  provider     = "mysql"
  url          = env("DATABASE_URL")
  relationMode = "prisma"
}

generator client {
  provider = "prisma-client-js"
}

model User {
  id           Int      @id @default(autoincrement())
  email        String   @unique
  name         String
  followers    Int      @default(0)
  isActive     Boolean  @default(true)
  registeredAt DateTime @default(now())
}

model Item {
  id    Int    @id @default(autoincrement())
  name  String
  price Float
}
```

```.env
DATABASE_URL='mysql://_RandomStringUserName_:_LongRandomStringPassword_@aws.connect.psdb.cloud/planetdb?sslaccept=strict'
```
My only concern is that PlanetScale recommends using ```db push``` instead of ```migrate dev```.

Does this help us in debugging the underlying issue with ```migrate dev``` not able to connect to localhost DB server ?

**Reply by janpio** [reply, marked as answer]
If `migrate dev` does not work locally for you, then `db push` also does not work. Just run it to confirm.

This means that the problem is with your local database (or the connection string you are trying to use for it). It is not a general Prisma problem when talking to a MySQL database.

**Reply by ManuWeb3** [reply]
Yes, just checked. ```db push``` failed locally.
Weird. 

Having noted that there are no issues with Prisma but the local DB, can you provide me any other insights/pointers to debug further at my end?

I already reached out to MySQL open community (when I wasn't even sure that issue lies with DB) and opened a discussion but never heard back.
I've already spent days trying to debug it.
Anything which you feel can help?

**Reply by janpio** [reply]
The difference between uppercase and lowercase database seems weird, and could indicate that something off is happening.
Besides that, nothing comes to mind, sorry.

**Reply by ManuWeb3** [reply]
Ok.
But, have upper/lower case been the issue, don't you think migration should have run with ```nextapp``` name?
Because, ```nextapp``` did appear everywhere (MySQL CLI, DataGrip IDE, etc.) in lowercase only unlike ```DBSampleDB``` name.

I know you already provided me with the best you could think of, just that this crossed my mind and I shared.

**Reply by janpio** [reply]
Yes, that also confuses me - if you really tried all combinations, this should exclude this as the source. But it is the only weird and non expected thing that we saw from your description and screenshots. It's not normal, so this is where I would start and figure out how to remove it from the equation. If the problem then still persists, you would know that it was not _that_.

**Reply by ManuWeb3** [reply]
It's indeed baffling.

Just 1 last thing for my clarity only: 
What did you refer to by **_it_** when you said "how to remove it from the equation" ?

Is it that I should work with PlanetScale or similar DB solutions rather than spending more time on debugging local DB issue?
Or something else?

**Reply by janpio** [reply]
The fact that your database somehow sometimes shows the database name with capitalization, and sometimes all lowercase. That should not be the case, `dbsampledb` and `DBSampleDB` should be two completely separate databases.

**Reply by ManuWeb3** [reply]
Ok.
Well, thanks Jan for all the supprt you provided.
Appreciate your patience :)

**Reply by Shivanand0297** [reply]
https://forum.codewithmosh.com/t/trouble-with-prisma-mysql-connection-string-when-using-socket-connection-method/23823

this helped me make sure that your password in the connection string is URL encoded.

**Comment by MrVhek** [comment]
Not sure if this would help, but you can try using localhost instead of 127.0.0.1, I already had issue with this.

**Reply by ManuWeb3** [reply]
Hey @MrVhek , thanks for the reply. Actually, I already tried this switch during my initial debugging but it did not help.
Either way, the issue still persisted.
I also tried to change the default port 3306 to 3309 in MySQL Server's Configurator and changed the port number in the connection string.
Still, the same issue.

**Comment by ghost** [comment]
I am also having this issue

**Reply by laxmariappan** [reply]
I got the same error, too 

> Error: P1001: Can't reach database server at `db`:`10010`
> 
> Please make sure your database server is running at `db`:`10010`.

`mysql://user:passwd@localhost:10010/dbname`

**Reply by ManuWeb3** [reply]
Mine never worked till date and now I've stopped worrying about it, already wasted tons of my time on it.
Kindly refer to (https://github.com/prisma/prisma/discussions/21666#discussioncomment-7462322) 
I'm happy working with PlanetScale's MYSQL dbms solution.

You may also want to give it a shot. Good luck.

**Comment by mconfjp** [comment]
I have a similar situation and I'm writing about it here.

# Situation
I was creating an app with Docker-compose.yml looking like this

```
version: '3.7'
services:
  app:
    build: .
    tty: true
    ports:
      - 3005:3005
      - 5555:5555
    volumes:
      - .:/app
      - /app/node_modules
    depends_on:
      - db
  db:
    image: mysql:8.0
    volumes:
      - db-store:/var/lib/mysql
      - ./logs:/var/log/mysql
      - ./docker/mysql/my.cnf:/etc/mysql/conf.d/my.cnf
      - ./docker/init:/docker-entrypoint-initdb.d
    environment:
      - MYSQL_DATABASE=${DB_NAME}
      - MYSQL_USER=${DB_USER}
      - MYSQL_PASSWORD=${DB_PASS}
      - MYSQL_ROOT_PASSWORD=${DB_PASS}
      - TZ=${TZ}
    ports:
      - ${DB_PORT}:3306
```

And the DATABASE _URL is this

```
DATABASE_URL="mysql://root:password@localhost:3306/database"
```

And same error

# Solution
DATABASE _URL was like this
```
DATABASE_URL="mysql://root:password@db:3306/database"
```

``prisma generate`` command location was in the app, so it seems that specifying localhost referred to the app itself.

**Reply by dominictobias** [reply]
Damn thanks. I don't know how `db` crept in there instead of `localhost`... trying stupid AI "vibe coding" for the first time and it's totally ret***ed

**Comment by Shivanand0297** [comment]
https://forum.codewithmosh.com/t/trouble-with-prisma-mysql-connection-string-when-using-socket-connection-method/23823

this helped me make sure your password in the connection string is URL encoded.

**Reply by Womoemy** [reply]
Using the percent encoder here also worked for me.

**Reply by pavanpandya** [reply]
Thanks for sharing this. I used the password generator, and it had a few wildcard symbols that were causing this issue.

**Comment by Tejas73** [comment]
Hey try this command because it worked for me when I faced the same issue today
`npx prisma migrate dev --name init  `

**Comment by babalomajiyezi** [comment]
I had the same error and to fix it I ran Docker engine in the background and ran `npx prisma migrate dev` .

**Comment by AndoRjh** [comment]
i have the same error too with my database on vercel, I don't know what happens exactly because there are a time that it works correctly and then after I have the same problem, I think that it is a bug of the prisma adapter but I'm searching the solution

**Comment by Coderamrin** [comment]
I had the same issue. 
fixed by adding timeout on the URL
Read this:
https://www.prisma.io/docs/orm/overview/databases/neon#resolving-connection-timeouts

**Comment by TheFabi8A** [comment]
I tried all the solutions given on stack overflow or in the official documentation of neon, prisma, etc. for a whole day. The only thing that magically worked was unchecking the "Internet Protocol version 6 (TCP/IPv6)" box in the Ethernet properties.

![image](https://github.com/user-attachments/assets/6871b8b0-cff9-4e07-860c-9cf1acbcace5)

**Comment by DanielLesko28** [comment]
Hey guys.
I had the same issue for a while and tried everything to set prisma with supabase correctly and actually this article helped me and I finally was able to connect to DB.

https://medium.com/@tibetdelek/supabase-and-prisma-setup-bfcd5d3c7e8b

**Comment by ofirgaash1** [comment]
solved for me by switching to more stable and fast network

**Comment by umapathikrishnank** [comment]
Faced the same P1001 issue  - the DB was reachable, but Prisma kept failing. I came here looking for answers, but after restarting my laptop it started working for some reason
